### PR TITLE
Add PUT endpoint to change room hosts

### PR DESCRIPTION
**Pull request by alankbi**
### List of Changes:

* Add PUT endpoint for changing room host
* Create new error type for invalid permissions
* Change room's set of users to list of users
* Add tests for new endpoint

### Notes:

* The set of users was changed to a list because sets don't have a get method, and we need that to ensure our objects are equal on both an application level and database level
* A new PUT request in Postman has been added for changing room hosts
* This PR has a lot of lines changed, but probably more than 75% of that is tests

### CODE REVIEW

**Comment by alankbi** [comment]
[Create PUT endpoint for changing the room host](https://trello.com/c/5ePjlDbw/137-create-put-endpoint-for-changing-the-room-host)

**Review by helenli8181 — APPROVED** [review]
This looks awesome! I just had a small comment but other than that it's good by me.

File: src/main/java/com/rocketden/main/dto/room/UpdateHostRequest.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+import com.rocketden.main.dto.user.UserDto;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class UpdateHostRequest {
+    private UserDto initiator;
+    private UserDto newHost;
```
Comment (zpChris) [inline src/main/java/com/rocketden/main/dto/room/UpdateHostRequest.java:11]: Can the host set themselves as the new host? I would assume yes, but want to make sure that if someone does that there aren't any unexpected consequences (bugs in the backend or whatnot for whatever reason).

File: src/test/java/com/rocketden/main/socket/RoomSocketTests.java [hunk 2]
```diff
@@ -98,24 +99,64 @@ public void handleFrame(StompHeaders headers, Object payload) {
                 blockingQueue.add((RoomDto) payload);
             }
         });
+    }
+
+    @Test
+    public void socketReceivesMessageOnJoin() throws Exception {
```
Comment (zpChris) [inline src/test/java/com/rocketden/main/socket/RoomSocketTests.java:105]: Thanks for adding this! This is unrelated to your current PR, right? This is a great test to add either way.

File: src/test/java/com/rocketden/main/api/RoomTests.java [hunk 3]
```diff
@@ -315,4 +313,163 @@ public void createAndJoinRoomNoUser() throws Exception {
...
+
+        jsonResponse = result.getResponse().getContentAsString();
+        actual = Utility.toObject(jsonResponse, ApiErrorResponse.class);
+
+        assertEquals(ERROR.getResponse(), actual);
+
+        // New host does not exist in the room
+        updateHostRequest = new UpdateHostRequest();
+        updateHostRequest.setInitiator(host);
+        updateHostRequest.setNewHost(new UserDto());
```
Comment (zpChris) [inline src/test/java/com/rocketden/main/api/RoomTests.java:460]: If I understand correctly, you are testing multiple different errors in the same method. Thoughts on taking these out and separating them into separate methods?

If you are concerned about the large amount of duplicate code beforehand, maybe place that in a setup helper function, and have a separate test (you may already) to ensure that that code is indeed working. Let me know your thoughts.

**Review by zpChris — CHANGES_REQUESTED** [review]
I left a few nits, requesting changes primarily because there is a test that I think could be separate into multiple different tests for readability and debugging purposes.

File: src/test/java/com/rocketden/main/entity/RoomEntityTests.java [hunk 4]
```diff
@@ -42,16 +42,41 @@ public void removeUserFromRoomSet() {
 
         User user = new User();
         user.setNickname("test");
+        user.setId(1);
         room.addUser(user);
         
         User userToRemove = new User();
-
         userToRemove.setNickname("nonexistent");
+        userToRemove.setId(2);
```
[the remark this answers is not in the document]
Comment (alankbi) [reply src/test/java/com/rocketden/main/entity/RoomEntityTests.java:50]: We talked about this in a call and I think reached a consensus - feel free to re-request if there's anything else that comes up tho

File: src/main/java/com/rocketden/main/dto/room/UpdateHostRequest.java (hunk 1 above)
Comment (alankbi) [reply src/main/java/com/rocketden/main/dto/room/UpdateHostRequest.java:11]: Just tested and yeah they can technically set themselves as the host (no errors occur, although I'm thinking in the frontend we can probably just prevent that from being an option)

File: src/test/java/com/rocketden/main/api/RoomTests.java (hunk 3 above)
Comment (alankbi) [reply src/test/java/com/rocketden/main/api/RoomTests.java]: Yeah that's a good idea, done!

File: src/main/java/com/rocketden/main/controller/v1/RoomController.java [hunk 5]
```diff
@@ -39,4 +41,10 @@ public RoomController(RoomService service) {
     public ResponseEntity<RoomDto> createRoom(@RequestBody CreateRoomRequest request) {
         return new ResponseEntity<>(service.createRoom(request), HttpStatus.CREATED);
     }
+
+    @PutMapping("/rooms/{roomId}/host")
+    public ResponseEntity<RoomDto> updateRoomHost(@PathVariable String roomId,
+                                                  @RequestBody UpdateHostRequest request) {
```
[the remark this answers is not in the document]
Comment (alankbi) [reply src/main/java/com/rocketden/main/controller/v1/RoomController.java:47]: Yeah based on what we talked about in standup (it might have been a while back) I think maybe we should change the join room and maybe get room endpoints to also use the path variables. I added a card for that in the backlog, but if we end up deciding to stick with room id as part of the request I can also come back and fix this if that sounds good

File: src/main/java/com/rocketden/main/dto/room/UpdateHostRequest.java (hunk 1 above)
Comment (zpChris) [reply src/main/java/com/rocketden/main/dto/room/UpdateHostRequest.java:11]: Ok awesome, thanks for verifying that, and yeah the frontend design sounds good.

File: src/main/java/com/rocketden/main/model/Room.java [hunk 6]
```diff
@@ -62,7 +62,11 @@ public boolean removeUser(User user) {
      * "equal" a user in the room at the application level, they are not the same at
      * the database level (different primary keys). This method therefore returns the
      * correct user object that's equal at both the application and database levels.
+     *
+     * @param userToFind the desired user to find
+     * @return the equivalent user in the database
      */
+
```
Comment (zpChris) [inline src/main/java/com/rocketden/main/model/Room.java:69]: nit: Remove line? Clearly you added that in, if there was a reason please lmk, but as I understood convent was to have no lines between the method comment and the method itself.

File: src/main/java/com/rocketden/main/model/Room.java [hunk 7]
```diff
@@ -62,7 +62,11 @@ public boolean removeUser(User user) {
      * "equal" a user in the room at the application level, they are not the same at
      * the database level (different primary keys). This method therefore returns the
      * correct user object that's equal at both the application and database levels.
+     *
+     * @param userToFind the desired user to find
+     * @return the equivalent user in the database
```
Comment (zpChris) [inline src/main/java/com/rocketden/main/model/Room.java:67]: Include the fact that you return `null` if that user is not in the database.

File: src/test/java/com/rocketden/main/api/RoomTests.java [hunk 8]
```diff
@@ -425,51 +372,116 @@ public void changeRoomHostFailure() throws Exception {
...
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .content(Utility.convertObjectToJsonString(joinRequest)))
+                .andDo(print()).andExpect(status().isOk())
+                .andReturn();
+
+        jsonResponse = result.getResponse().getContentAsString();
+        room = Utility.toObject(jsonResponse, RoomDto.class);
+
+        assertEquals(host, room.getHost());
+        assertEquals(2, room.getUsers().size());
```
Comment (zpChris) [inline src/test/java/com/rocketden/main/api/RoomTests.java:483]: nit, personal preference: Verify that the non-host user is indeed the one that you added.

**Review by zpChris — APPROVED** [review]
LGTM! I left a few nits for quick fixes, but after that, the PR looks great!

File: src/main/java/com/rocketden/main/model/Room.java (hunk 6 above)
Comment (alankbi) [reply src/main/java/com/rocketden/main/model/Room.java]: Haha yep you're right, accidentally left a new line there

File: src/test/java/com/rocketden/main/api/RoomTests.java (hunk 8 above)
Comment (alankbi) [reply src/test/java/com/rocketden/main/api/RoomTests.java:484]: Added a contains statement to check if the list contains the UserDto that we specified
